cvat/apps/consensus/consensus_reports.py

```diff
@@ -0,0 +1,493 @@
+    comparison_summary: ComparisonReportComparisonSummary
+    frame_results: Dict[int, ComparisonReportFrameSummary]
+
+    @property
+    def conflicts(self) -> List[AnnotationConflict]:
+        return list(itertools.chain.from_iterable(r.conflicts for r in self.frame_results.values()))
+
+    @property
+    def consensus_score(self) -> int:
+        mean_consensus_score = 0
+        frame_count = 0
+        for frame_result in self.frame_results.values():
+            if not isinstance(frame_result.consensus_score, NoneType):
+                mean_consensus_score += frame_result.consensus_score
+                frame_count += 1
+
+        return np.round(100 * (mean_consensus_score / (frame_count or 1)))
+
+    def _fields_dict(self, *, include_properties: Optional[List[str]] = None) -> dict:
+        return super()._fields_dict(
+            include_properties=include_properties
+            or [
+                "consensus_score",
+            ]
+        )
+
+    @classmethod
+    def from_dict(cls, d: Dict[str, Any]) -> ComparisonReport:
+        return cls(
+            parameters=ComparisonParameters.from_dict(d["parameters"]),
+            comparison_summary=ComparisonReportComparisonSummary.from_dict(d["comparison_summary"]),
+            frame_results={
+                int(k): ComparisonReportFrameSummary.from_dict(v)
+                for k, v in d["frame_results"].items()
+            },
+        )
+
+    def to_json(self) -> str:
+        d = self.to_dict()
+
+        # String keys are needed for json dumping
+        d["frame_results"] = {str(k): v for k, v in d["frame_results"].items()}
+        return dump_json(d).decode()
+
+    @classmethod
+    def from_json(cls, data: str) -> ComparisonReport:
+        return cls.from_dict(parse_json(data))
+
+
+def generate_job_consensus_report(
+    consensus_settings: ConsensusSettings,
+    errors,
+    consensus_job_data_providers: List[JobDataProvider],
+    merged_dataset: dm.Dataset,
+    merger: IntersectMerge,
+    assignees: List[User],
+    assignee_report_data: Dict[User, Dict[str, float]],
+) -> ComparisonReport:
+
+    frame_results: Dict[int, ComparisonReportFrameSummary] = {}
+    frames = set()
+    conflicts_count = len(errors)
+    frame_wise_conflicts: Dict[int, List[AnnotationConflict]] = {}
+    frame_wise_consensus_score: Dict[int, List[float]] = {}
+    conflicts: List[AnnotationConflict] = []
+
+    for error in errors:
+        error_type = str(type(error)).split(".")[-1].split("'")[0]
+        error_type = ConsensusConflictType[error_type].value
+        annotation_ids = []
+        error_annotations = []
+
+        for arg in error.args:
+            if isinstance(arg, Annotation):
+                error_annotations.append(arg)
+
+        for annotation in error_annotations:
+            # the annotation belongs to which consensus dataset
+            idx = merger._dataset_map[merger._item_map[merger._ann_map[id(annotation)][1]][1]][1]
```

Comment:
There should be a function for this in the merger class.

Reply:
```py
    def add_item_error(self, error, *args, **kwargs):
        self.errors.append(error(self._item_id, *args, **kwargs))
```
There's only this function which handles error, but it doesn't do all that we want...

Reply:
I can see there is `get_ann_source()`, which does half of the work. But you can always define your own function in the class, if there is no fitting one. It will be better than accessing private fields.

Reply:
Ah, I see. I will create a similar function which does rest of the work, instead of accessing the private fields.

Reply:
```py
idx = merger._dataset_map[merger._item_map[merger._ann_map[id(annotation)][1]][1]][1]
annotation_ids.append(consensus_job_data_providers[idx].dm_ann_to_ann_id(annotation))
```
By any chance do you have any clue why `consensus_job_data_providers[idx].dm_ann_to_ann_id(annotation)` would be failing because of keyerror, i.e., the annotation isn't a valid annotation for the job data provider I am using...?

Reply:
No, it's hard to say it without the full context. Here are 2 possible reasons:
- it's a nested point in a skeleton, which wasn't added to the index
- `idx` is not the index of the source provider

Reply:
It was the 2nd one. It was happening because the `self.error` list in `IntersectMerge` was storing information about all the errors encountered since it's initialisation.